crates/core/src/operations/update.rs

```diff
@@ -413,6 +439,48 @@ async fn execute(
         serde_json::to_value(&metrics)?,
     );
 
+    match tracker.collect().await {
+        Ok(batches) => {
+            if !batches.is_empty() {
+                debug!(
+                    "Collected {} batches to write as part of this transaction:",
+                    batches.len()
+                );
+                let config = WriterConfig::new(
+                    batches[0].schema().clone(),
+                    snapshot.metadata().partition_columns.clone(),
+                    writer_properties.clone(),
+                    None,
+                    None,
+                );
+
+                let store = Arc::new(PrefixStore::new(
+                    log_store.object_store().clone(),
+                    "_change_data",
+                ));
+                let mut writer = DeltaWriter::new(store, config);
+                for batch in batches {
+                    writer.write(&batch).await?;
+                }
+                // Add the AddCDCFile actions that exist to the commit
+                actions.extend(writer.close().await?.into_iter().map(|add| {
+                    Action::Cdc(AddCDCFile {
+                        // This is a gnarly hack, but the action needs the nested path, not the
```

Comment:
Can't we pass the write_action_type to the writer_config and then use some generic return types, so that based on the writer_config we either return Action::Cdc or Action::Add?

Reply:
`writer_config` is already a mess, I do not want to add more of anything there. Since I started this pull request we added another field there :upside_down_face: 

I wouldn't call this hack a hack though! That makes it sounds so ugly :laughing:   The concern I have with making the DeltaWriter generic with regards to the Action generated is that the Add has some more expansive requirements  with file stats. 

One option I considered was adding an `From<Add>` on `AddCDCFile` to at least relegate this code into a better location.  I'm not happy with our writer structure at the moment anyways, and it's due for some surgery but I don't want to do that surgery here :frowning: `